File: packages/javascript/src/__tests__/publisher/topic-publisher.spec.ts
import { TopicPublisher } from '../../publisher/topic-publisher';
import connection from '../../core/connection';
import config from '../../core/config';

// Mock dependencies
jest.mock('../../core/connection');
jest.mock('nats', () => ({
  headers: jest.fn(() => ({
    set: jest.fn(),
  })),
}));

describe('TopicPublisher', () => {
  let publisher: TopicPublisher;
  let mockJetstream: any;
  let mockHeaders: any;

  beforeEach(() => {
    publisher = new TopicPublisher();

    mockHeaders = {
      set: jest.fn(),
    };

    mockJetstream = {
      publish: jest.fn().mockResolvedValue(undefined),
    };

    (connection.ensureConnection as jest.Mock) = jest.fn().mockResolvedValue(undefined);
    (connection.getJetStream as jest.Mock) = jest.fn().mockReturnValue(mockJetstream);

    const { headers } = require('nats');
    (headers as jest.Mock).mockReturnValue(mockHeaders);

    config.configure({
      natsUrls: 'nats://localhost:4222',
      env: 'test',
      appName: 'test-app',
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('publishToTopic', () => {
    it('should publish to a simple topic', async () => {
      const message = { text: 'Hello', user_id: 123 };

      await publisher.publishToTopic('notifications', message);

      expect(connection.ensureConnection).toHaveBeenCalled();
      expect(connection.getJetStream).toHaveBeenCalled();
      expect(mockJetstream.publish).toHaveBeenCalledWith(
        'test.test-app.notifications',
        expect.any(String),
        expect.objectContaining({
          msgID: expect.any(String),
          headers: mockHeaders,
        })
      );
    });

    it('should publish to a hierarchical topic', async () => {
      const message = { to: 'user@example.com', subject: 'Welcome' };

      await publisher.publishToTopic('notifications.email', message);

      const publishCall = mockJetstream.publish.mock.calls[0];
      expect(publishCall[0]).toBe('test.test-app.notifications.email');
    });

    it('should publish to multi-level hierarchical topic', async () => {
      const message = { order_id: 456, total: 99.99 };

      await publisher.publishToTopic('order.processing.completed', message);

      const publishCall = mockJetstream.publish.mock.calls[0];
      expect(publishCall[0]).toBe('test.test-app.order.processing.completed');
    });

    it('should create topic envelope with correct structure', async () => {
      const message = { text: 'Hello', user_id: 123 };
      const options = {
        event_id: 'custom-event-id',
        trace_id: 'trace-123',
        occurred_at: new Date('2025-01-01T00:00:00Z'),
        message_type: 'urgent',
      };

      await publisher.publishToTopic('notifications', message, options);

      const publishCall = mockJetstream.publish.mock.calls[0];
      const envelopeJson = publishCall[1];
      const envelope = JSON.parse(envelopeJson);

      expect(envelope).toEqual({
        event_id: 'custom-event-id',
        schema_version: 1,
        topic: 'notifications',
        message_type: 'urgent',
        producer: 'test-app',
        occurred_at: '2025-01-01T00:00:00.000Z',
        trace_id: 'trace-123',
        message,
      });
    });

    it('should generate event_id if not provided', async () => {
      const message = { text: 'Hello' };

      await publisher.publishToTopic('notifications', message);

      const publishCall = mockJetstream.publish.mock.calls[0];
      const envelopeJson = publishCall[1];
      const envelope = JSON.parse(envelopeJson);

      expect(envelope.event_id).toMatch(
        /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i
      );
    });

    it('should set topic header', async () => {
      const message = { text: 'Hello' };

      await publisher.publishToTopic('notifications', message);

      expect(mockHeaders.set).toHaveBeenCalledWith('topic', 'notifications');
    });

    it('should set trace_id header when provided', async () => {
      const message = { text: 'Hello' };
      const options = { trace_id: 'trace-123' };

      await publisher.publishToTopic('notifications', message, options);

      expect(mockHeaders.set).toHaveBeenCalledWith('trace_id', 'trace-123');
    });

    it('should handle publish errors', async () => {
      const error = new Error('Publish failed');
      mockJetstream.publish.mockRejectedValueOnce(error);

      const message = { text: 'Hello' };

      await expect(publisher.publishToTopic('notifications', message)).rejects.toThrow(
        'Publish failed'
      );
    });

    it('should normalize topic names correctly', async () => {
      const message = { text: 'Hello' };

      // Test uppercase
      await publisher.publishToTopic('NOTIFICATIONS', message);
      expect(mockJetstream.publish.mock.calls[0][0]).toBe('test.test-app.notifications');

      jest.clearAllMocks();

      // Test special characters
      await publisher.publishToTopic('user@events!', message);
      expect(mockJetstream.publish.mock.calls[0][0]).toBe('test.test-app.user_events_');
    });

    it('should preserve dots in hierarchical topics', async () => {
      const message = { text: 'Hello' };

      await publisher.publishToTopic('analytics.user.signup', message);

      expect(mockJetstream.publish.mock.calls[0][0]).toBe('test.test-app.analytics.user.signup');
    });

    it('should preserve wildcards in topic names', async () => {
      const message = { text: 'Hello' };

      await publisher.publishToTopic('notifications.*', message);

      expect(mockJetstream.publish.mock.calls[0][0]).toBe('test.test-app.notifications.*');
    });
  });

  describe('publishToTopics', () => {
    it('should publish to multiple topics', async () => {
      const message = { action: 'user_login', user_id: 123 };
      const topics = ['notifications', 'audit.user_events'];

      const results = await publisher.publishToTopics(topics, message);

      expect(mockJetstream.publish).toHaveBeenCalledTimes(2);
      expect(results).toEqual({
        notifications: true,
        'audit.user_events': true,
      });
    });

    it('should handle partial failures', async () => {
      mockJetstream.publish
        .mockResolvedValueOnce(undefined)
        .mockRejectedValueOnce(new Error('Failed'));

      const message = { text: 'Hello' };
      const topics = ['topic1', 'topic2'];

      const results = await publisher.publishToTopics(topics, message);

      expect(results).toEqual({
        topic1: true,
        topic2: false,
      });
    });

    it('should publish the same message to all topics', async () => {
      const message = { action: 'password_changed', user_id: 123 };
      const topics = ['notifications', 'audit.events', 'security.alerts'];

      await publisher.publishToTopics(topics, message);

      expect(mockJetstream.publish).toHaveBeenCalledTimes(3);

      // Check that all envelopes have the same message
      for (let i = 0; i < 3; i++) {
        const envelopeJson = mockJetstream.publish.mock.calls[i][1];
        const envelope = JSON.parse(envelopeJson);
        expect(envelope.message).toEqual(message);
      }
    });
  });

  describe('publish (domain/resource/action)', () => {
    it('should map domain/resource/action to topic', async () => {
      const payload = { id: '123', name: 'John' };

      await publisher.publish('users', 'user', 'created', payload);

      const publishCall = mockJetstream.publish.mock.calls[0];
      const subject = publishCall[0];
      const envelopeJson = publishCall[1];
      const envelope = JSON.parse(envelopeJson);

      // Should map to topic: users.user.created
      expect(subject).toBe('test.test-app.users.user.created');
      expect(envelope.topic).toBe('users.user.created');
      expect(envelope.domain).toBe('users');
      expect(envelope.resource).toBe('user');
      expect(envelope.action).toBe('created');
    });

    it('should include domain/resource/action fields in envelope', async () => {
      const payload = { id: '123', name: 'John' };

      await publisher.publish('orders', 'order', 'placed', payload);

      const publishCall = mockJetstream.publish.mock.calls[0];
      const envelopeJson = publishCall[1];
      const envelope = JSON.parse(envelopeJson);

      expect(envelope).toMatchObject({
        topic: 'orders.order.placed',
        domain: 'orders',
        resource: 'order',
        action: 'placed',
        resource_id: '123',
        message: payload,
      });
    });
  });

  describe('cross-environment support', () => {
    it('should work with different environments', async () => {
      config.configure({
        env: 'production',
        appName: 'prod-app',
      });

      const message = { text: 'Hello' };
      await publisher.publishToTopic('notifications', message);

      const publishCall = mockJetstream.publish.mock.calls[0];
      expect(publishCall[0]).toBe('production.prod-app.notifications');

      const envelopeJson = publishCall[1];
      const envelope = JSON.parse(envelopeJson);
      expect(envelope.producer).toBe('prod-app');
    });
  });
});

File: packages/ruby/spec/topic_publisher_spec.rb
# frozen_string_literal: true

require 'nats_pubsub/topic_publisher'

RSpec.describe NatsPubsub::TopicPublisher do
  let(:publisher) { described_class.new }
  let(:mock_jts) { double('JetStream') }
  let(:mock_config) { double('Config', env: 'development', app_name: 'app', use_outbox: false) }

  before do
    allow(NatsPubsub).to receive(:config).and_return(mock_config)
    allow(NatsPubsub::Connection).to receive(:connect!).and_return(mock_jts)
    allow(NatsPubsub::Logging).to receive(:info)
    allow(NatsPubsub::Logging).to receive(:error)
  end

  describe '#publish_to_topic' do
    let(:topic) { 'notifications' }
    let(:message) { { type: 'email', to: 'user@example.com' } }

    before do
      allow(mock_jts).to receive(:publish).and_return(double(duplicate?: false, error: nil))
    end

    it 'publishes message to topic' do
      expect(mock_jts).to receive(:publish).with(
        'development.app.notifications',
        kind_of(String),
        hash_including(header: kind_of(Hash))
      )
      publisher.publish_to_topic(topic, message)
    end

    it 'includes topic in envelope' do
      expect(mock_jts).to receive(:publish) do |_subject, payload, _options|
        envelope = Oj.load(payload)
        expect(envelope['topic']).to eq('notifications')
      end
      publisher.publish_to_topic(topic, message)
    end

    it 'includes message in envelope' do
      expect(mock_jts).to receive(:publish) do |_subject, payload, _options|
        envelope = Oj.load(payload)
        expect(envelope['message']).to eq(message.stringify_keys)
      end
      publisher.publish_to_topic(topic, message)
    end

    it 'generates event_id' do
      expect(mock_jts).to receive(:publish) do |_subject, payload, _options|
        envelope = Oj.load(payload)
        expect(envelope['event_id']).to be_a(String)
        expect(envelope['event_id']).not_to be_empty
      end
      publisher.publish_to_topic(topic, message)
    end

    it 'accepts custom event_id' do
      custom_id = 'custom-event-123'
      expect(mock_jts).to receive(:publish) do |_subject, payload, _options|
        envelope = Oj.load(payload)
        expect(envelope['event_id']).to eq(custom_id)
      end
      publisher.publish_to_topic(topic, message, event_id: custom_id)
    end

    it 'includes trace_id when provided' do
      trace_id = 'trace-123'
      expect(mock_jts).to receive(:publish) do |_subject, payload, _options|
        envelope = Oj.load(payload)
        expect(envelope['trace_id']).to eq(trace_id)
      end
      publisher.publish_to_topic(topic, message, trace_id: trace_id)
    end

    it 'normalizes topic name' do
      expect(mock_jts).to receive(:publish).with(
        'development.app.my_topic_',
        anything,
        anything
      )
      publisher.publish_to_topic('My Topic!', message)
    end

    it 'returns true on success' do
      result = publisher.publish_to_topic(topic, message)
      expect(result).to be true
    end

    it 'returns false on error' do
      allow(mock_jts).to receive(:publish).and_raise(StandardError)
      result = publisher.publish_to_topic(topic, message)
      expect(result).to be false
    end
  end

  describe '#publish_to_topics' do
    let(:topics) { %w[notifications audit] }
    let(:message) { { action: 'user_login' } }

    before do
      allow(mock_jts).to receive(:publish).and_return(double(duplicate?: false, error: nil))
    end

    it 'publishes to multiple topics' do
      expect(mock_jts).to receive(:publish).with(
        'development.app.notifications',
        anything,
        anything
      )
      expect(mock_jts).to receive(:publish).with(
        'development.app.audit',
        anything,
        anything
      )
      publisher.publish_to_topics(topics, message)
    end

    it 'returns hash with results' do
      results = publisher.publish_to_topics(topics, message)
      expect(results).to be_a(Hash)
      expect(results['notifications']).to be true
      expect(results['audit']).to be true
    end

    it 'handles partial failures' do
      allow(mock_jts).to receive(:publish) do |subject, _payload, _options|
        raise StandardError if subject.include?('audit')

        double(duplicate?: false, error: nil)
      end

      results = publisher.publish_to_topics(topics, message)
      expect(results['notifications']).to be true
      expect(results['audit']).to be false
    end
  end

  describe 'envelope structure' do
    let(:topic) { 'test_topic' }
    let(:message) { { key: 'value' } }

    before do
      allow(mock_jts).to receive(:publish).and_return(double(duplicate?: false, error: nil))
    end

    it 'includes schema_version' do
      expect(mock_jts).to receive(:publish) do |_subject, payload, _options|
        envelope = Oj.load(payload)
        expect(envelope['schema_version']).to eq(1)
      end
      publisher.publish_to_topic(topic, message)
    end

    it 'includes producer' do
      expect(mock_jts).to receive(:publish) do |_subject, payload, _options|
        envelope = Oj.load(payload)
        expect(envelope['producer']).to eq('app')
      end
      publisher.publish_to_topic(topic, message)
    end

    it 'includes occurred_at' do
      expect(mock_jts).to receive(:publish) do |_subject, payload, _options|
        envelope = Oj.load(payload)
        expect(envelope['occurred_at']).to be_a(String)
        expect(envelope['occurred_at']).to match(/\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}/)
      end
      publisher.publish_to_topic(topic, message)
    end

    it 'accepts custom occurred_at' do
      custom_time = Time.utc(2024, 1, 1, 12, 0, 0)
      expect(mock_jts).to receive(:publish) do |_subject, payload, _options|
        envelope = Oj.load(payload)
        expect(envelope['occurred_at']).to eq('2024-01-01T12:00:00Z')
      end
      publisher.publish_to_topic(topic, message, occurred_at: custom_time)
    end
  end
end
